tests/integration/test_charm.py

```diff
@@ -61,3 +62,47 @@ async def test_with_ingress(
     )
     assert response.status_code == 200
     assert "Welcome to the Matrix" in response.text
+
+
+async def test_server_name_changed(
+    model: Model, synapse_app: Application, different_server_name: str
+):
+    """
+    arrange: build and deploy the Synapse charm.
+    act: change server_name via juju config.
+    assert: the Synapse application should prevent the change.
+    """
+    await model.applications[synapse_app.name].set_config({"server_name": different_server_name})
+    await model.wait_for_idle()
+    unit = model.applications[synapse_app.name].units[0]
+    # Status string defined in Juju
+    # https://github.com/juju/juju/blob/2.9/core/status/status.go#L150
+    assert unit.workload_status == "blocked"
+    assert "is different from the existing" in unit.workload_status_message
+
+
+async def test_reset_instance_action(
+    model: Model, synapse_app: Application, different_server_name: str
+):
+    """
+    arrange: build and deploy the Synapse charm.
+    act: change server_name via juju config.
+    assert: the old instance is deleted and the new one configured.
```

Comment:
Minor nitpick, how about using a fixture here? Another synapse charm fixture with a "function" scope might be useful :D 
I.e.
```python3
@pytest_asyncio.fixture(scope="function", name="name_configured_app")
async def name_configured_app_fixture(synapse_app: Application):
    await synapse_app.set_config({"server_name":different_server_name})
    
    yield synapse_app
    
    await synaps_app.reset_config(["server_name"])
```

This will make the tests flow a bit more focused as:
```python3
"""
arrange: given a configured synapse application.
act: when reset instance action is triggered.
assert: the old instance is deleted and the new one configured.
"""
```

Reply:
Nice suggestion, thanks!